### ISSUE TITLE
ModelCheckpoint: manual-optimization pre-step snapshot is not applied when the save is deferred to on_validation_end

### ISSUE THREAD

**Issue by ErenAta16** [Lightning-AI/pytorch-lightning#21842]
### Bug description

`ModelCheckpoint` with manual optimization, `every_n_train_steps`, and a validation-only monitored metric (e.g. `monitor="val_loss"` logged in `validation_step`) saves the checkpoint credited as "best" with the **live, fully-trained, post-optimization weights at the end of the epoch**, not the pre-optimization weights captured for the step that actually produced the best score. This defeats the entire point of the pre-optimization-state-snapshot mechanism added in #21239.

`on_train_batch_end` (`src/lightning/pytorch/callbacks/model_checkpoint.py`, manual-optimization branch) correctly swaps in `pl_module.saved_models[latest_step]` before saving and restores the live weights afterward, when the monitored metric is already available:

```python
# lines 381-397
with torch.no_grad():
    original_state = {k: v.detach().clone() for k, v in pl_module.layer.state_dict().items()}
    try:
        saved_state = saved_models[latest_step]
        pl_module.layer.load_state_dict(saved_state)
        self._save_topk_checkpoint(trainer, monitor_candidates)
        self._save_last_checkpoint(trainer, monitor_candidates)
        self._last_time_checked = time.monotonic()
    finally:
        pl_module.layer.load_state_dict(original_state)
```

But when the monitored metric isn't available yet (the normal case for a validation-only metric during training), the save is deferred instead (line 361-363):

```python
if self.monitor is not None and self.monitor not in monitor_candidates:
    self._defer_save_until_validation = True
    return
```

And the deferred path, `on_validation_end` (lines 507-511), does the save with **no swap at all**:

```python
if self._defer_save_until_validation:
    self._save_topk_checkpoint(trainer, monitor_candidates)
    self._save_last_checkpoint(trainer, monitor_candidates)
    self._defer_save_until_validation = False
    return
```

This calls straight into `_save_checkpoint` -> `trainer.save_checkpoint(...)`, which just dumps whatever the live model state is at that moment, i.e. after every `optimizer.step()` call for the whole epoch (or however many steps ran before validation), not the snapshot corresponding to the step that produced the best monitored value.

`_save_checkpoint`'s own docstring claims: *"For manual optimization, we rely on the fact that the model's training_step method saves the model state before the optimizer step, so we can use that state directly"*, but that guarantee is only actually implemented at the one call site inside `on_train_batch_end`, not in `on_validation_end`.

This is a regression of #20947 ("the best checkpointed model is the model obtained after backpropagation and not the one used for computing the loss"), reintroduced by an incomplete fix in #21239. #21239's own test (`test_model_checkpoint_manual_opt`) only monitors a training-step metric (`monitor="loss"`), which is always present in `on_train_batch_end`'s `monitor_candidates` and therefore never exercises the `_defer_save_until_validation` branch at all, so it never caught that the swap logic doesn't exist in `on_validation_end`.

### How to reproduce the bug

```python
import torch
from torch.utils.data import DataLoader, Dataset
from lightning.pytorch import LightningModule, Trainer
from lightning.pytorch.callbacks import ModelCheckpoint


class FakeDataset(Dataset):
    def __init__(self, n=4):
        self.data = [torch.randn(3) for _ in range(n)]
        self.labels = [torch.randint(0, 2, (1,)) for _ in range(n)]

    def __len__(self):
        return len(self.data)

    def __getitem__(self, idx):
        return self.data[idx], self.labels[idx]


class SimpleModule(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(3, 1)
        self.automatic_optimization = False
        self.saved_models = {}
        self.fake_val_losses = [1.0, 1.0, 0.0, 1.0]  # batch_idx=2 is "best"

    def training_step(self, batch, batch_idx):
        out = self.layer(batch[0])
        loss = torch.nn.functional.binary_cross_entropy_with_logits(out, batch[1].float())
        # documented pattern from the library's own warning message: snapshot BEFORE optimizer.step()
        self.saved_models[batch_idx] = {k: v.detach().clone() for k, v in self.layer.state_dict().items()}
        opt = self.optimizers()
        opt.zero_grad()
        self.manual_backward(loss)
        opt.step()
        return loss

    def validation_step(self, batch, batch_idx):
        self.log("val_loss", torch.tensor(self.fake_val_losses[batch_idx]), on_epoch=True)

    def configure_optimizers(self):
        return torch.optim.SGD(self.parameters(), lr=0.5)


model = SimpleModule()
trainer = Trainer(
    max_epochs=1,
    callbacks=[ModelCheckpoint(
        monitor="val_loss", mode="min", save_top_k=1, every_n_train_steps=1,
        every_n_epochs=0, save_on_train_epoch_end=False, save_weights_only=True, save_last=False,
    )],
    num_sanity_val_steps=0,
)
trainer.fit(model, DataLoader(FakeDataset(), batch_size=1), DataLoader(FakeDataset(), batch_size=1))

best_ckpt = torch.load(trainer.checkpoint_callback.best_model_path, weights_only=True)["state_dict"]

def matches(saved_state, ckpt):
    return all(torch.equal(saved_state[name.removeprefix("layer.")], val) for name, val in ckpt.items())

any_pre_opt_match = any(matches(s, best_ckpt) for s in model.saved_models.values())
live_weights = {k: v.detach().clone() for k, v in model.layer.state_dict().items()}
matches_live = all(torch.equal(live_weights[name.removeprefix("layer.")], val) for name, val in best_ckpt.items())

print("Checkpoint matches ANY pre-optimization snapshot (expected):", any_pre_opt_match)
print("Checkpoint matches the LIVE, fully-trained end-of-epoch weights instead:", matches_live)
```

Actual output:
```
Checkpoint matches ANY pre-optimization snapshot (expected): False
Checkpoint matches the LIVE, fully-trained end-of-epoch weights instead: True
```

The saved "best" checkpoint doesn't match any of the four pre-optimization snapshots captured during the epoch (it should match the one from the step with the best `val_loss`), it matches the fully-trained, post-all-optimizer-steps weights instead.

### What version are you seeing the problem on?

master, v2.6

### How to reproduce the bug

(see script above)

### Environment

- PyTorch Lightning Version: 2.6.5 (pip install lightning), confirmed the same logic is present on current `master` by reading `src/lightning/pytorch/callbacks/model_checkpoint.py` directly
- PyTorch Version: 2.x, CPU only, no GPU needed to reproduce
- Python version: 3.12.10
- OS: Windows 11
- How installed: pip

### More info

Related: #20947 (the original bug), #21239 (the fix that introduced this regression by only covering the non-deferred `on_train_batch_end` path). A fix would need to apply the same pre-optimization-state swap in `on_validation_end`'s `_defer_save_until_validation` branch that already exists in `on_train_batch_end`.

**Comment by deependujha** [comment]
Hi @ErenAta16, thanks for raising the issue.

However, I'm not sure if it is actually an issue. We don't run `training step0 -> validation step0 -> training step1 -> validation step1` and so on.
The flow is `training step 0 -> trianing step 1-> ...` then `validation step 0 -> validation step 1 -> ...`.

I'm talking about the general case when you've not specified how often validation runs. So, during validation, irrespective of which validation step produced the best loss, we will save live weights during validation only.

I modified your reproducible code minimally only, and it very evidently shows what's going on.

```python
import torch
from torch.utils.data import DataLoader, Dataset
from lightning.pytorch import LightningModule, Trainer
from lightning.pytorch.callbacks import ModelCheckpoint


class FakeDataset(Dataset):
    def __init__(self, n=4):
        self.data = [torch.randn(3) for _ in range(n)]
        self.labels = [torch.randint(0, 2, (1,)) for _ in range(n)]

    def __len__(self):
        return len(self.data)

    def __getitem__(self, idx):
        return self.data[idx], self.labels[idx]


class SimpleModule(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(3, 1)
        self.automatic_optimization = False
        self.saved_models = {}
        self.fake_val_losses = [1.0, 1.0, 0.0, 1.0]  # batch_idx=2 is "best"

    def training_step(self, batch, batch_idx):
        print(f"training step")
        out = self.layer(batch[0])
        loss = torch.nn.functional.binary_cross_entropy_with_logits(out, batch[1].float())
        # documented pattern from the library's own warning message: snapshot BEFORE optimizer.step()
        self.saved_models[batch_idx] = {k: v.detach().clone() for k, v in self.layer.state_dict().items()}
        opt = self.optimizers()
        opt.zero_grad()
        self.manual_backward(loss)
        opt.step()
        return loss

    def validation_step(self, batch, batch_idx):
        print(f"validation step")
        self.log("val_loss", torch.tensor(self.fake_val_losses[batch_idx]), on_epoch=True)

    def configure_optimizers(self):
        return torch.optim.SGD(self.parameters(), lr=0.5)


model = SimpleModule()
trainer = Trainer(
    accelerator="cpu",
    max_epochs=1,
    callbacks=[ModelCheckpoint(
        monitor="val_loss", mode="min", save_top_k=1, every_n_train_steps=1,
        every_n_epochs=0, save_on_train_epoch_end=False, save_weights_only=True, save_last=False,
    )],
    num_sanity_val_steps=0,
)
trainer.fit(model, DataLoader(FakeDataset(), batch_size=1), DataLoader(FakeDataset(), batch_size=1))

best_ckpt = torch.load(trainer.checkpoint_callback.best_model_path, weights_only=True)["state_dict"]

def matches(saved_state, ckpt):
    return all(torch.equal(saved_state[name.removeprefix("layer.")], val) for name, val in ckpt.items())

any_pre_opt_match = any(matches(s, best_ckpt) for s in model.saved_models.values())
live_weights = {k: v.detach().clone() for k, v in model.layer.state_dict().items()}
matches_live = all(torch.equal(live_weights[name.removeprefix("layer.")], val) for name, val in best_ckpt.items())

print("Checkpoint matches ANY pre-optimization snapshot (expected):", any_pre_opt_match)
print("Checkpoint matches the LIVE, fully-trained end-of-epoch weights instead:", matches_live)
```

and the output is look for is:

```
training step (wo -> w1)
training step (w1 -> w2)
training step (w2 -> w3)
training step (w3 -> w4)
validation step (w4)
validation step (w4)
validation step (w4)
validation step (w4)
```

since all the validation is running with `w4` that's why it isn't present in the `saved_weights`, and nothing seems wrong to me, as we don't validate with pre-optimized weights typically, if you need that use `training time metric` which produce it on the basis of live weights in training step.

ps: The issue was closed due to slip of hands. Sorry.

**Comment by deependujha** [comment]
Following up on my previous comment, one part of this report is valid, and there's a bigger problem sitting right next to it.

`_save_checkpoint`'s docstring claims unconditionally that for manual optimization we rely on the training_step snapshot. That's only true at the one call site in `on_train_batch_end`, and it's what led to this report. Worth scoping the docstring and adding a comment in the deferred branch explaining why no swap happens there. Happy to take that.

**Separate, larger problem in the same code path.**

While re-reading the manual-opt branch in `on_train_batch_end`, the swap is gated on `hasattr(pl_module, "layer") and isinstance(pl_module.layer, torch.nn.Module)`, with an adjacent comment noting that the *test case* stores state in
`model.saved_models`. A few consequences:

- Any module that **`doesn't`** happen to name its submodule `layer` (`self.backbone`,
  `self.encoder`, ...) falls through to the `else` branch and gets the 15-line
  `rank_zero_warn` on **every** train batch.
- When it does fire, only `pl_module.layer.state_dict()` is swapped, but `trainer.save_checkpoint()` serializes the whole module — so for any model with more than one submodule the checkpoint mixes pre-step weights for `layer` with post-step weights for everything else. That state never existed during training.
- `latest_step = max(saved_models.keys())` uses the documented example's `batch_idx` keys, which reset each epoch, so from epoch 2 onward it can restore a snapshot written in a previous epoch.

The second and third points aren't documentation issues.

**Comment by ErenAta16** [comment]
You're right about the flow — validation runs as a block after training, not interleaved, so the framing in my report was wrong. Thanks for the correction.

I checked your three follow-up points against `master`. Two hold exactly as stated; the third needs one qualification.

**The `layer`-only swap.** Confirmed. Lines 383-396 clone and restore `pl_module.layer.state_dict()`, but `_save_topk_checkpoint` → `trainer.save_checkpoint()` serialises the whole module. Any model with a second submodule gets pre-step weights for `layer` and post-step weights for everything else, a combination that never existed during training.

**`latest_step = max(saved_models.keys())`.** Confirmed, and the documented example makes it concrete. That example keys by `batch_idx` and prunes with `del self.saved_models[batch_idx - 10]`, so it keeps a 10-entry window that survives the epoch boundary:

```
epoch 1 end: keys=[90..99]                    max()=99 -> snapshot@e1b99
epoch 2, 3 batches in: keys=[0,1,2,90..99]    max()=99 -> snapshot@e1b99
```

From epoch 2 onward every save restores an epoch-1 snapshot until `batch_idx` climbs back past 99.

**The warning firing on every batch.** The gate is real — `hasattr(pl_module, "layer")` at line 371 hardcodes the submodule name from the test case, and the comment on line 366 says so outright, so `self.backbone` or `self.encoder` falls through. On the frequency, `rank_zero_warn` is a plain `warnings.warn` with `stacklevel=4` and no cache, so under Python's default filter it prints once per location rather than once per batch. It does repeat whenever filters are set to `always`, which is common under pytest. The sharper version of the point is that `lightning_utilities` ships `WarningCache` precisely for this and this call site doesn't use it.

On my side: I'm closing #21843. It moves the existing swap into a shared helper and calls it from the deferred path too — which means it would apply the `max(keys)` selection and the `layer`-only restore at a second call site, propagating all three defects rather than fixing them. That's the wrong shape given what's above.

Happy to pick up whichever piece you don't want once the design is settled — the `layer` gate and the snapshot-key scheme both look like they need an interface decision (some explicit hook or documented contract) rather than a patch.
